[Static Analyzer CI] Refactor FindUnexpectedStaticAnalyzerResults to be more modular and understandable

briannafan opened the pull request:
#### b9aa0a838ddc32b4740ccec0cef523204083d841
<pre>
[Static Analyzer CI] Refactor FindUnexpectedStaticAnalyzerResults to be more modular and understandable
<a href="https://bugs.webkit.org/show_bug.cgi?id=287184">https://bugs.webkit.org/show_bug.cgi?id=287184</a>
<a href="https://rdar.apple.com/144336563">rdar://144336563</a>

Reviewed by Aakash Jain.

Separate logic into FindUnexpectedStaticAnalyzerResults and FindUnexpectedStaticAnalyzerResultsWithoutChange.
This removes the need for the confusing `use_expectations` argument.

* Tools/CISupport/ews-build/steps.py:
(ScanBuild.addResultsSteps):
(ScanBuildWithoutChange.addResultsSteps):
(FindUnexpectedStaticAnalyzerResults.__init__): Remove use_expectations.
(FindUnexpectedStaticAnalyzerResults.run): Remove logic that will be used in WithoutChange.
(FindUnexpectedStaticAnalyzerResults.decode_results_data): Added. To be reused in multiple locations.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db): Use decode_results_data.

(FindUnexpectedStaticAnalyzerResultsWithoutChange): Move all the logic for the second run here.
(FindUnexpectedStaticAnalyzerResultsWithoutChange.run):

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/290412@main">https://commits.webkit.org/290412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/115a787bb2c2944083684379fce0b633b8b265aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89952 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44851 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/40725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92004 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/9868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17764 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/94952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/40725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92953 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/9868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/9868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39859 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/9868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/37024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/17140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/96776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/89537 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/17396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14135 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/17150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/16891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/20343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7161,31 +7160,10 @@ def run(self):
-                ]
-        else:
-            self.command += ['--check-expectations']
+        self.command += ['--build-output', SCAN_BUILD_OUTPUT_DIR, '--check-expectations']
```
Do we need it in separate line or can merge it in `self.command =` ?

briannafan replied:
It can be merged!

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7364,6 +7343,57 @@ def getResultSummary(self):
+    name = 'find-unexpected-results-without-change'
+
+    @defer.inlineCallbacks
+    def run(self):
```
Please check if we can share more code especially for run() by moving it to base class entirely, and have run() call a method to calculate the self.command which is overridden in derived class. For eg, see `setLayoutTestCommand()` in this file.

If that's doesn't work, no worries.

briannafan replied:
This may be difficult given the subtle differences in logic in what needs to be parsed after running the command, but I will explore it in the next PR.

briannafan committed:
[Static Analyzer CI] Refactor FindUnexpectedStaticAnalyzerResults to be more modular and understandable
https://bugs.webkit.org/show_bug.cgi?id=287184
rdar://144336563

Reviewed by Aakash Jain.

Separate logic into FindUnexpectedStaticAnalyzerResults and FindUnexpectedStaticAnalyzerResultsWithoutChange.
This removes the need for the confusing `use_expectations` argument.

* Tools/CISupport/ews-build/steps.py:
(ScanBuild.addResultsSteps):
(ScanBuildWithoutChange.addResultsSteps):
(FindUnexpectedStaticAnalyzerResults.__init__): Remove use_expectations.
(FindUnexpectedStaticAnalyzerResults.run): Remove logic that will be used in WithoutChange.
(FindUnexpectedStaticAnalyzerResults.decode_results_data): Added. To be reused in multiple locations.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db): Use decode_results_data.

(FindUnexpectedStaticAnalyzerResultsWithoutChange): Move all the logic for the second run here.
(FindUnexpectedStaticAnalyzerResultsWithoutChange.run):

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/290412@main
